04-cs-fundamentals/internship/hash-tables.md

```diff
@@ -0,0 +1,319 @@
+
+`h(k) = floor(m * (k * A mod 1))`
+
+So if:
+
+- `k` = 37
+- `A` = 0.31
+- `m` = 100
+
+`h(37) = floor(100 * (37 * 0.31 % 1))`
+
+`h(37) = floor(100 * (11.47 % 1))`
+
+`h(37) = floor(100 * (0.47000000000000064))`
+
+`h(37) = floor(47.000000000000064)`
+
+
+`h(37) = 47`
+
+**Exercise** Assuming `A` is 0.71 and `m` = 100, answer the following
+
+<details>
+  <summary> if k is 15, what is h(k) using the multiplication method?  What about the division method?
+  </summary>
+  Multiplication method: 64 <br />
+  Division method: 15
+</details>
+
+<br />
+
+You are **not** expected to memorize these methods, but rather be able to see how they work to identify a bucket to find values in a hash table.
+
+There are **many** different ways to write a hash function.  There is not a mathematically provable perfect solution, in the general case.  Instead developers use statistical analysis with experimental data to judge the performance of a given hashing function.  When evaluating a hashing function developers want a function which is easy to compute, maps any given input to valid buckets, and minimizes collisions.
+
+[Ruby uses the MurmurHash](https://sites.google.com/site/murmurhash/) hashing function internally.
+
+## Collision Handling
+
+All general-purpose hashing functions will encounter collisions.  When two keys are mapped to the same bucket something has to happen to manage it.  We will look at three methods: chaining, linear probing, quadratic probing and rehashing.
+
+### Chaining
+
+The first solution, as we discussed in the classroom portion of Ada, is to make each bucket of the hash table's internal array the head of a linked list.  Linked lists are quick to insert and remove items O(1) and can store an arbitrary number of elements.  Unfortunately if there are a high number of items in the same bucket, finding an element starts to approach O(n).  However if the hashing function does a good job of spreading elements out over different buckets the linked lists will be small and the time to search for an item in the hash table approaches O(1).  
+
+![Chaining](../classroom/images/Example-hash.png)
+
+### Linear Probing
+
+Chaining required a secondary data structure in the array.  Linear probling is instead rather simple.  When inserting a new element into the hash table, the hash function will return with an index number.  If the bucket at that index number is occupied (there's a collision), then the hash table checks the next element in the array, and the next and so on until an empty bucket is found.  
+
+![Linear Probing](images/linear-probing.png)
+
+Similar to chaining, searching for an empty bucket has a worst-case of O(n).  Also, because we are looking only at adjacent elements for empty buckets, this can lead to a problem with _clustering_.  Clustering means that elements are grouping together into one area in the hash table.  
+
+![Clustering](images/linear-probing-clustering2.png)
```

Comment:
It looks like you only use `linear-probing-clustering2.png`.  Consider removing `linear-probing-clustering.png` and renaming this one.